### PR TITLE
tests: lib: mem_alloc: Add tests to validate dynamic allocation

### PR DESCRIPTION
**Pull request by pswarnak**
Add new test cases to validate dynamic memory allocation
functions such as malloc, calloc, realloc using minimal libc
and newlibc implementation of standard C library.

Signed-off-by: Praful Swarnakar <praful.swarnakar@intel.com>

### CODE REVIEW

File: tests/lib/mem_alloc/src/main.c [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	char *ptr = NULL;
+
+#ifdef CONFIG_NEWLIB_LIBC
+	void *heap_base = NULL;
+
+	z_newlib_get_heap_bounds(&heap_base, &max_heap_size);
+#else
+	max_heap_size = MAX_HEAP_SIZE;
+#endif
+	ptr = (char *) malloc(max_heap_size + 1);
```
Comment (andrewboie) [inline tests/lib/mem_alloc/src/main.c]: char * cast shouldn't be necessary, malloc returns void

**Comment by andrewboie** [comment]
Some CI failures:


```
starting test - test_memalloc_max



    Assertion failed at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/lib/mem_alloc/src/main.c:220: test_memalloc_max: (errno not equal to ENOMEM
```

It seems newlib is not setting errno when a malloc() call fails.
BTW, the Nios II targets force newlib to be enabled so thats why qemu_nios2 fails for minimal libc scenario.

File: tests/lib/mem_alloc/src/main.c [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	zassert_is_null(ptr, "malloc passed unexpectedly");
+	zassert_equal(errno, ENOMEM, "Wrong error number, errno: %d", errno);
```
Comment (andrewboie) [inline tests/lib/mem_alloc/src/main.c]: remove this check.
malloc() setting errno is in the C99 spec, but not in C11 spec, and newlib currenty is not setting errno in this situation.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
Remove ENOMEM check due to it not being well-specified/supported, and blacklist this test for native_posix as native_posix seems to be using host c library

File: tests/lib/mem_alloc/src/main.c (hunk 1 above)
Comment (pswarnak) [reply tests/lib/mem_alloc/src/main.c]: yeah, C performs implicit conversion. Updated, please check. Thanks.

File: tests/lib/mem_alloc/src/main.c (hunk 2 above)
Comment (pswarnak) [reply tests/lib/mem_alloc/src/main.c]: Sure, i did refer minimal_libc malloc implementation that sets errno.
Removed to allow test pass for both minimal and newlib implementation.
Thanks for the information.

**Comment by pswarnak** [comment]
@andrewboie Thank you for the review comments. All comments incorporated, shippable checks passed successfully. Kindly check.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**